File: src/core/refs/ref-manager.ts
import { Repository } from '@/core/repo';
import { FileUtils, logger } from '@/utils';
import path from 'path';
import fs from 'fs-extra';

/**
 * RefManager handles Git references (refs) - the human-readable names for commits.
 *
 * Git References Overview:
 * - Refs are pointers to commits (SHA-1 hashes)
 * - Stored as simple text files containing commit SHAs
 * - Located in .git/refs/ directory
 *
 * Reference Types:
 * 1. **HEAD**: Points to the current branch or commit
 * 2. **Branches**: refs/heads/<branch-name>
 * 3. **Tags**: refs/tags/<tag-name>
 * 4. **Remote branches**: refs/remotes/<remote>/<branch>
 *
 * Reference Resolution:
 * - Symbolic refs: "ref: refs/heads/master" (HEAD pointing to a branch)
 * - Direct refs: "abc123..." (direct SHA-1 hash)
 *
 * File Structure:
 * .git/
 * ├── HEAD                    # Current branch/commit
 * ├── refs/
 * │   ├── heads/             # Local branches
 * │   │   ├── master         # Contains SHA of master's tip
 * │   │   └── feature-x      # Contains SHA of feature-x's tip
 * │   └── tags/              # Tags
 * │       └── v1.0.0         # Contains SHA of tagged commit
 */
export class RefManager {
  private refsPath: string;
  private headPath: string;

  public static readonly REFS_DIRNAME = 'refs' as const;
  public static readonly SYMBOLIC_REF_PREFIX = 'ref: ' as const;
  public static readonly HEAD_FILE = 'HEAD' as const;

  constructor(repository: Repository) {
    const gitDir = repository.gitDirectory().fullpath();
    this.refsPath = path.join(gitDir, RefManager.REFS_DIRNAME);
    this.headPath = path.join(gitDir, RefManager.HEAD_FILE);
  }

  /**
   * Get the full path for the refs directory
   */
  public getRefsPath(): string {
    return this.refsPath;
  }

  /**
   * Initialize the ref manager
   */
  public async init() {
    await FileUtils.createDirectories(this.refsPath);
    await fs.writeFile(this.headPath, 'ref: refs/heads/master\n', 'utf8');
  }

  /**
   * Read a reference and return its content
   */
  public async readRef(ref: string): Promise<string> {
    try {
      const fullPath = this.resolveReferencePath(ref);

      if (!(await FileUtils.exists(fullPath))) {
        throw new Error(`Ref ${ref} not found`);
      }

      const content = await fs.readFile(fullPath, 'utf8');
      return content.trim();
    } catch (error) {
      throw new Error(`Error reading ref ${ref}: ${error}`);
    }
  }

  /**
   * Update a reference with a new SHA-1 hash
   */
  public async updateRef(refPath: string, sha: string): Promise<void> {
    const fullPath = this.resolveReferencePath(refPath);
    await FileUtils.createDirectories(path.dirname(fullPath));
    await fs.writeFile(fullPath, sha + '\n', 'utf8');
    logger.info(`Updated ref ${refPath} to ${sha}`);
  }

  /**
   * Resolve a reference to its final SHA-1 hash
   */
  public async resolveReferenceToSha(refPath: string): Promise<string> {
    const maxDepth = 10;
    let currentRef = refPath;

    for (let depth = 0; depth < maxDepth; depth++) {
      let content: string;
      try {
        content = await this.readRef(currentRef);
      } catch (error) {
        throw new Error(`Error reading ref ${currentRef}: ${error}`);
      }

      if (content.startsWith(RefManager.SYMBOLIC_REF_PREFIX)) {
        currentRef = content.substring(RefManager.SYMBOLIC_REF_PREFIX.length);
        continue;
      }

      if (this.isSha1(content)) return content;
    }

    throw new Error(`Reference depth exceeded for ${refPath}`);
  }

  /**
   * Delete a reference
   */
  public async deleteRef(ref: string): Promise<boolean> {
    const fullPath = this.resolveReferencePath(ref);

    if (!(await FileUtils.exists(fullPath))) {
      return false;
    }

    try {
      await fs.unlink(fullPath);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Get the full path for a HEAD reference
   */
  public getHeadPath(): string {
    return this.headPath;
  }

  /**
   * Check if a reference exists
   */
  public async exists(ref: string): Promise<boolean> {
    return await FileUtils.exists(this.resolveReferencePath(ref));
  }

  /**
   * Get the full path for a reference
   */
  private resolveReferencePath(refInput: string): string {
    const ref = refInput.trim();

    if (ref === RefManager.HEAD_FILE) {
      return this.headPath;
    }

    // If ref starts with "refs/", don't duplicate the refs root
    if (ref.startsWith(`${RefManager.REFS_DIRNAME}/`)) {
      return path.join(this.refsPath, ref.slice(RefManager.REFS_DIRNAME.length + 1));
    }

    return path.join(this.refsPath, ref);
  }

  /**
   * Check if a string is a valid SHA-1 hash
   */
  private isSha1(str: string): boolean {
    return /^[0-9a-f]{40}$/i.test(str);
  }
}

File: sourcecontrol/pkg/repository/refs/ref_manager.go
package refs

import (
	"fmt"
	"strings"

	"github.com/utkarsh5026/SourceControl/pkg/common/fileops"
	"github.com/utkarsh5026/SourceControl/pkg/objects"
	"github.com/utkarsh5026/SourceControl/pkg/repository/scpath"
	"github.com/utkarsh5026/SourceControl/pkg/repository/sourcerepo"
)

const (
	// SymbolicRefPrefix is the prefix for symbolic references
	SymbolicRefPrefix = "ref: "

	// MaxRefDepth is the maximum depth for resolving symbolic references
	MaxRefDepth = 10
)

// RefManager handles Git references (refs) - human-readable names for commits
type RefManager struct {
	refsPath scpath.SourcePath
	headPath scpath.SourcePath
}

// NewRefManager creates a new reference manager for the given repository
func NewRefManager(repo sourcerepo.Repository) *RefManager {
	sourceDir := repo.SourceDirectory()
	return &RefManager{
		refsPath: sourceDir.RefsPath(),
		headPath: sourceDir.HeadPath(),
	}
}

// Init initializes the ref manager by creating the refs directory and HEAD file
func (rm *RefManager) Init() error {
	if err := fileops.EnsureDir(rm.refsPath.ToAbsolutePath()); err != nil {
		return fmt.Errorf("failed to create refs directory: %w", err)
	}

	defaultRef := "ref: refs/heads/master\n"
	if err := fileops.WriteConfigString(rm.headPath.ToAbsolutePath(), defaultRef); err != nil {
		return fmt.Errorf("failed to create HEAD file: %w", err)
	}

	return nil
}

// ReadRef reads a reference and returns its content
func (rm *RefManager) ReadRef(ref RefPath) (string, error) {
	fullPath := rm.resolveReferencePath(ref)

	content, err := fileops.ReadStringStrict(fullPath.ToAbsolutePath())
	if err != nil {
		return "", fmt.Errorf("error reading ref %s: %w", ref, err)
	}

	return content, nil
}

// UpdateRef updates a reference with a new SHA-1 hash
func (rm *RefManager) UpdateRef(ref RefPath, hash objects.ObjectHash) error {
	// Validate the hash
	if err := hash.Validate(); err != nil {
		return fmt.Errorf("invalid hash: %w", err)
	}

	fullPath := rm.resolveReferencePath(ref).ToAbsolutePath()

	if err := fileops.EnsureParentDir(fullPath); err != nil {
		return fmt.Errorf("failed to create ref directory: %w", err)
	}

	content := hash.String() + "\n"
	if err := fileops.WriteConfigString(fullPath, content); err != nil {
		return fmt.Errorf("failed to write ref %s: %w", ref, err)
	}

	return nil
}

// ResolveToSHA resolves a reference to its final SHA-1 hash, following symbolic refs
func (rm *RefManager) ResolveToSHA(ref RefPath) (objects.ObjectHash, error) {
	currentRef := ref

	for range MaxRefDepth {
		content, err := rm.ReadRef(currentRef)
		if err != nil {
			return "", fmt.Errorf("error reading ref %s: %w", currentRef, err)
		}

		if after, ok := strings.CutPrefix(content, SymbolicRefPrefix); ok {
			target := after
			currentRef = RefPath(target)
			continue
		}

		hash, err := objects.NewObjectHashFromString(content)
		if err != nil {
			return "", fmt.Errorf("invalid ref content: %s", content)
		}

		return hash, nil
	}

	return "", fmt.Errorf("reference depth exceeded for %s", ref)
}

// DeleteRef deletes a reference
func (rm *RefManager) DeleteRef(ref RefPath) (bool, error) {
	fullPath := rm.resolveReferencePath(ref).ToAbsolutePath()

	exists, err := fileops.Exists(fullPath)
	if err != nil {
		return false, err
	}
	if !exists {
		return false, nil
	}

	if err := fileops.SafeRemove(fullPath); err != nil {
		return false, err
	}

	return true, nil
}

// Exists checks if a reference exists
func (rm *RefManager) Exists(ref RefPath) (bool, error) {
	fullPath := rm.resolveReferencePath(ref).ToAbsolutePath()
	return fileops.Exists(fullPath)
}

// GetHeadPath returns the full path to the HEAD file
func (rm *RefManager) GetHeadPath() scpath.SourcePath {
	return rm.headPath
}

// GetRefsPath returns the full path to the refs directory
func (rm *RefManager) GetRefsPath() scpath.SourcePath {
	return rm.refsPath
}

// resolveReferencePath resolves a RefPath to its full filesystem path
func (rm *RefManager) resolveReferencePath(ref RefPath) scpath.SourcePath {
	refStr := strings.TrimSpace(ref.String())

	if refStr == scpath.HeadFile {
		return rm.headPath
	}

	if after, ok := strings.CutPrefix(refStr, scpath.RefsDir+"/"); ok {
		relPath := after
		return rm.refsPath.Join(relPath)
	}

	return rm.refsPath.Join(refStr)
}
